include/alpaka/dev/DevOacc.hpp

```diff
@@ -138,15 +135,32 @@ namespace alpaka
             };
         } // namespace detail
     } // namespace oacc
+
     //! The OpenACC device handle.
     class DevOacc
         : public concepts::Implements<ConceptCurrentThreadWaitFor, DevOacc>
         , public concepts::Implements<ConceptDev, DevOacc>
     {
         friend struct traits::GetDevByIdx<PltfOacc>;
 
+        static auto& device(int iDevice)
+        {
+            static std::vector<std::shared_ptr<oacc::detail::DevOaccImpl>> devices(getDevCount<PltfOacc>());
+            static std::mutex mutex;
+
+            auto& d = devices.at(static_cast<unsigned>(iDevice));
+            if(!d)
+            {
+                std::lock_guard<std::mutex> lock(mutex);
+                if(!d)
+                    return devices[static_cast<unsigned>(iDevice)]
+                        = std::make_shared<oacc::detail::DevOaccImpl>(iDevice);
+            }
+            return d;
```

Comment:
2189d98 changes it to the `unique_ptr` implementation. I can still look into adding an atomic store for the assignment for platforms where pointers are written in multiple pieces if you think this is a potential issue.

Reply:
First of, there are no sequence points any more since C++11. You are right that there is a happens before relationship between the construction of the `unique_ptr` and the assignment of the address into the `vector`. However, that sequencing concerns visible side effects within a single thread. And the optimizer can reorder the construction and assignment, because it does not matter within a single thread.

In a multithreaded scenario you also need synchronization operations such as mutex locks/unlocks or atomic loads/stores with acquire/release semantic. This is necessary because also the CPU can freely reorder the instruction stream. The unlock of your mutex is a release operation, so your changes should become visible. However, any other thread reading `d` misses an acquire operation., because release operations only synchronize with acquire operations and not ordinary reads. So theoretically, there is a data race.
See also this blog post: https://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/

I am tempted to believe this code can work in practice because the initial value of `d` is `nullptr`, visible to all threads and if a thread missreads an already initialized `d` as `nullptr` (e.g. from cache) it would then acquire the lock, which is an acquire operation synchronizing with a potential release operation from a different thread, and then have a valid read.

Anyway, I have argued enough and I'll let you proceed with what you think is best.

Reply:
> I am tempted to believe this code can work in practice because the initial value of d is nullptr, [...] synchronizing with a potential release operation from a different thread, and then have a valid read.

This is what I am arguing, yes.

> [...] However, any other thread reading d misses an acquire operation., because release operations only synchronize with acquire operations and not ordinary reads. So theoretically, there is a data race.

This is what we need to talk about, not the atomicity of the reads. Sorry I missed that. Using atomic load/store on the pointers is not so simple, however, as `std::unique_ptr` does not specialize atomic operations, so we would have to have a lock for each thread, every time. I just pushed c890ef61a8506a3ae63490836550cf848143fea5 which uses `std::atomic<T*>` which should do what you want, though in a (inevitably) long way.

> See also this blog post: https://preshing.com/20130930/double-checked-locking-is-fixed-in-cpp11/

This one suggests `std::atomic_thread_fence`, before giving the pure atomic solution as more elegant, which it is not in this case. So using `std::atomic_thread_fence` with `std::unique_ptr` and no atomics should be simplest solution. This would be fine, except if pointers are written piecemeal, which would require atomics to be solved. The post for some (maybe this) reason used the fence `and` atomics.

So.. what do you think? If you vote for  c890ef61a8506a3ae63490836550cf848143fea5 then that `ManageAtomicPtrVector` would need to got into a separate header....

Reply:
@bernhardmgruber I guess an alternative to this would be to have a `std::vector<std::pair<std::once_flag, std::unique_ptr<..>>>` and use `std::call_once` for the initialization to let the STL handle the locking. Instead of the pair one could also have a small type that handles the call to `call_once` and only returns references to ready objects.

Reply:
I replaced it with the simpler `std::call_once`-verison now. There is a nested but generic helper class. As I do not see any other use cases for this in Alpaka at the moment, I suggest to keep it nested for now and merge as is.

@bernhardmgruber can you please re-review now, it is ready to merge from my end.

Reply:
... ready to be merged after rebase and squashing, that is..